app/api/items/save_grade.go

```diff
@@ -0,0 +1,275 @@
+	}
+	service.MustNotBeError(err)
+
+	apiError := service.NoError
+	if apiError = checkSaveGradeTokenParams(user, &requestData); apiError != service.NoError {
+		return apiError
+	}
+
+	var validated, keyObtained bool
+	err = srv.Store.InTransaction(func(store *database.DataStore) error {
+		var hasAccess bool
+		var reason error
+		hasAccess, reason, err = store.Items().CheckSubmissionRights(requestData.TaskToken.Converted.LocalItemID, user)
+		service.MustNotBeError(err)
+
+		if !hasAccess {
+			apiError = service.ErrForbidden(reason)
+			return nil // commit! (CheckSubmissionRights() changes the DB sometimes)
+		}
+
+		validated, keyObtained = saveGradingResultsIntoDB(store, user, &requestData)
+		return nil
+	})
+	if apiError != service.NoError {
+		return apiError
+	}
+	service.MustNotBeError(err)
+
+	if validated && requestData.TaskToken.Converted.AccessSolutions != nil && !(*requestData.TaskToken.Converted.AccessSolutions) {
+		requestData.TaskToken.AccessSolutions = formdata.AnythingFromString(`"1"`)
+	}
+	requestData.TaskToken.PlatformName = srv.TokenConfig.PlatformName
+	newTaskToken, err := requestData.TaskToken.Sign(srv.TokenConfig.PrivateKey)
+	service.MustNotBeError(err)
+
+	service.MustNotBeError(render.Render(w, r, service.CreationSuccess(map[string]interface{}{
+		"task_token":   newTaskToken,
+		"bValidated":   validated,
+		"bKeyObtained": keyObtained,
+	})))
+	return service.NoError
+}
+
+func saveGradingResultsIntoDB(store *database.DataStore, user *database.User,
+	requestData *SaveGradeRequest) (validated, keyObtained bool) {
+	const todo = "todo"
+	score := requestData.ScoreToken.Converted.Score
+	userAnswerID := requestData.ScoreToken.Converted.UserAnswerID
+
+	// TODO: handle validation in a proper way (what did he mean??)
+	if score > 99 {
+		validated = true
+	}
+	service.MustNotBeError(store.UserAnswers().ByID(userAnswerID).
+		Where("idUser = ?", user.UserID).
+		Where("idItem = ?", requestData.TaskToken.Converted.LocalItemID).
+		UpdateColumn(map[string]interface{}{
+			"sGradingDate": gorm.Expr("NOW()"),
+			"bValidated":   validated,
+			"iScore":       score,
+		}).Error())
+	// Build query to update users_items
+	// The iScore is set towards the end, so that the IF condition on
+	// sBestAnswerDate is computed before iScore is updated
+	columnsToUpdate := []string{
+		"nbTasksTried",
+		"sLastActivityDate",
+		"sBestAnswerDate",
+		"sLastAnswerDate",
+		"iScore",
+	}
+	values := []interface{}{
+		1,
+		gorm.Expr("NOW()"),
+		gorm.Expr("IF(? > iScore, NOW(), sBestAnswerDate)", score),
+		gorm.Expr("NOW()"),
+		gorm.Expr("GREATEST(?, iScore)", score),
+	}
+	if validated {
+		// Item was validated
+		columnsToUpdate = append(columnsToUpdate,
+			"sAncestorsComputationState", "bValidated", "bKeyObtained", "sValidationDate",
+		)
+		values = append(values,
+			todo, 1, 1, gorm.Expr("IFNULL(sValidationDate, NOW())"))
+		keyObtained = true
+	} else {
+		// Item wasn't validated, check if we unlocked something
```

Comment:
I don't get why only unlocking if not validated.

Reply:
It's the case when iScoreMinUnlock <= 99

Reply:
The whole discussion around why this "else" has to be extremely well summarized/explained here.

Reply:
added a comment